draft-ietf-jsonpath-base.md

```diff
@@ -259,7 +259,7 @@ Parameter:
 : Formal parameter that can take Arguments (actual parameters) in a function-expression.
 
 Normalized Path:
-: A simple form of JSONPath expression that identifies a node in a value by
+: A minimal form of JSONPath expression that identifies a node in a value by
   providing a query that results in exactly that node.  Similar
```

Comment:
So let's do the "unique" here:

```suggestion
  providing a unique query that results in exactly that node.  Similar
```

Reply:
Here, "unique" doesn't work.  Identifying `foo` in `{"foo": "bar"}` can be achieved using `$['foo']` and `$.foo`.  The key here is that no shorthand syntaxes are used.

Secondly, are Normalized Paths necessarily Singular Paths?  If so, we should state that.

I think these two requirements (Singular Path and expressed without shorthand) are sufficient to define Normalized Path.

Reply:
> Here, "unique" doesn't work. Identifying `foo` in `{"foo": "bar"}` can be achieved using `$['foo']` and `$.foo`. The key here is that no shorthand syntaxes are used.

The way I read the spec is that the terminology section gives a loose definition of Normalized Path which is tightened up in section 2.9. With that reading, it would be ok to add uniqueness to the term definition because that falls out of the tighter definition in section 2.9. Maybe we need to add a forward reference from the term definition to section 2.9 to make it clear this is the way to read the spec? (I wouldn't want to add the full definition of Normalized Path to the terminology section. )

> 
> Secondly, are Normalized Paths necessarily Singular Paths? If so, we should state that.

Yes, they are. Section 2.9 already says:

>> Note: Normalized Paths are Singular Queries, but not all Singular Queries are Normalized Paths.

> 
> I think these two requirements (Singular Path and expressed without shorthand) are sufficient to define Normalized Path.

Unfortunately, those conditions are not sufficient. As section 2.9 says:

>> For example, $[-3] is a Singular Query, but is not a Normalized Path.

`$[-3]` is a Singular Query and is expressed without shorthand. (NB. the spec doesn't describe such array indices as shorthand.)

(PS. Pardon the above markup. I used `>>` for quotes from the draft to distinguish them from quotes of the earlier comment.)

Reply:
All of that is fair, I suppose.

The thing with `$[-3]` is that it's not unique _for all JSON instances_.  Processing that path on JSON arrays of different lengths will select nodes at different positions, but processing `$[3]` on those arrays will always select the fourth node (if it exists).

Maybe

> providing a unique query that results in exactly that node for all JSON instances.

(or "data" or whatever instead of "instances" [ref](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/issues/427))

Reply:
Ok, how about `$[1:2]` (as a synonym for `$[1]`)? It's a Singular Query, without shorthand, and results in the same node for all JSON instances.

Reply:
Yeah, I don't know.  This is hard to nail down, and it's late for me. 🙂

Reply:
There is nothing that is fundamentally better about always using an unsigned integer to represent normalized paths than using a negative integer.
We just decided always using the unsigned integer integer is more appealing (probably because we feel that this is a more fundamental way to access array elements, but that is not fundamentally so...).
So normalized paths are first and foremost a convention that lead to them being unique for each node.

Reply:
We could be more explicit and add a sentence:
Each node in an argument is identified by exactly one normalize path (we say, the normalized path is "unique" for that node), and, to be a normalized path for a specific argument, the normalize path needs to identify exactly one node.

Reply:
```suggestion
  providing a query that results in exactly that node.  Each node in an argument is identified by exactly one Normalized Path (we say, the Normalized Path is "unique" for that node), and, to be a Normalized Path for a specific argument, the Normalized Path needs to identify exactly one node. Similar
```